lockc/src/runc.rs

```diff
@@ -0,0 +1,589 @@
+    let bundle_path = container_bundle.as_ref();
+    let config_path = bundle_path.join("config.json");
+    let f = fs::File::open(config_path.clone())?;
+    let r = io::BufReader::new(f);
+
+    let config: ContainerConfig = serde_json::from_reader(r)?;
+
+    // Kubernetes
+    if let Some(annotations) = config.annotations {
+        debug!(
+            "detected kubernetes container with bundle {}, config {}",
+            bundle_path.display(),
+            config_path.display(),
+        );
+        // containerd
+        if annotations.contains_key(ANNOTATION_CONTAINERD_LOG_DIRECTORY) {
+            // containerd doesn't expose k8s namespaces directly. They have
+            // to be parsed from the log directory path, where the first
+            // part of the filename is the namespace.
+            let log_directory = &annotations[ANNOTATION_CONTAINERD_LOG_DIRECTORY];
+            debug!(
+                "detected k8s+containerd container with log directory {}",
+                log_directory
+            );
+            let log_path = std::path::PathBuf::from(log_directory);
+            let file_name = log_path.file_name().unwrap().to_str().unwrap();
+            let mut splitter = file_name.split('_');
+            let namespace = splitter.next().unwrap().to_string();
+
+            return Ok((ContainerType::KubernetesContainerd, namespace));
+        // containerd
+        } else if annotations.contains_key(ANNOTATION_CONTAINERD_SANDBOX_ID) {
+            // When a container is running as a part of a previously created
+            // pod, the log directory path has to be retrieved from the
+            // sandbox container.
+            let sandbox_id = &annotations[ANNOTATION_CONTAINERD_SANDBOX_ID];
+            debug!(
+                "detected k8s+containerd container with sandbox id {}",
+                sandbox_id
+            );
+
+            // Go one directory up from the current bundle.
+            let mut ancestors = bundle_path.ancestors();
+            ancestors.next();
+            if let Some(v) = ancestors.next() {
+                // Then go to sandbox_id directory (sandbox's bundle).
+                let new_bundle = v.join(sandbox_id);
+                return container_type_data(new_bundle);
+            }
+        }
+        // TODO(vadorovsky): Support more Kubernetes CRI implementations.
+        // They all come with their own annotations, so we will have to
+        // handle more keys here.
+    }
+
+    // Docker
+    for mount in config.mounts {
+        let source: Vec<&str> = mount.source.split('/').collect();
+        if source.len() > 1 && source[source.len() - 1] == "hostname" {
+            let config_v2 = str::replace(&mount.source, "hostname", "config.v2.json");
+            debug!("detected docker container with config path {}", config_v2);
+            return Ok((ContainerType::Docker, config_v2));
+        }
+    }
+
+    Ok((ContainerType::Unknown, String::from("")))
```

Comment:
Containing an `Option<String>` as value `None` instead of empty `String` could be better depending on your preference